### ISSUE TITLE
esl vs ssl

### ISSUE THREAD

**Issue by Lufei-github** [valeoai/ESL#1]
I am familar with your paper ADVENT and paper BDL's SSL, so I directly read your code. If my understanding is correct, the main
 
difference between esl and ssl  is in below code:

x = predicted_entropy[predicted_label==i]                         vs     x = predicted_prob[predicted_label==i]

thres[thres<cfg.ESL.THRESHOLD]=cfg.ESL.THRESHOLD     vs      thres[thres>0.9]=0.9

label[(entropy>thres[i])*(label==i)] = 255                           vs     label[(prob<thres[i])*(label==i)] = 255

Then I do a small experiment, I trained my unsupervised domain adaptation segmentation network, the obtained mIOU is 41.31, 

then I use the 41.31 mIOU 's model to run SSL.py, use the pseudo to train the target domain with a supervised way, then reach 

to 43.64 mIOU. Then, I modify the SSL mainly with upper code to ESL, to obtain the pseudo and train with target, however, the 

mIOU is 42.68. Besides, the SSL method mainly about 43, 43.2 mIOU, and ESL method mainly about 41.8, 42 mIOU, even 41.4 

mIOU. So I think its very strange for me. Do I do some wrong in modifying the code? Can you give me some suggestions?

**Comment by SportaXD** [comment]
The results you obtain are definitely very strange, the performance should be at least comparable to SSL.
Are you computing the entropy properly? What value are you using for the threshold? We use 0.1 as default value for the threshold in our code.
Maybe you can also compare visually the pseudo-labels between the two methods, a lot of the pixels should be shared.

**Comment by Lufei-github** [comment]
![ESL](https://user-images.githubusercontent.com/47933628/98761716-07de8680-2411-11eb-9e53-38a0e910eb48.png)

**Comment by Lufei-github** [comment]
Here is modified ESL, under your code’s instructions. I direct use your compute_entropy function and the 0.1 threshold. The blue code is my modification. Could you spend one minute in the upper code and tell me where is wrong? Thanks so much!

**Comment by SportaXD** [comment]
Your code looks correct... What are the thresholds you are getting with the median (you see them with `print(thres)` in your code)? Could you share some visualization of the pseudo-labels with both SSL and ESL on a couple of images? 
If everything is fine on the pseudo-label side, maybe you had an unlucky adversarial training that failed to get a good optimum; using pseudo-labels (SSL or ESL) usually helps in reducing the variance between experiments but those cases sometimes happen.

**Comment by Lufei-github** [comment]
![pseudolabel](https://user-images.githubusercontent.com/47933628/99021058-caa60000-259a-11eb-8af7-64c5702eb136.png)
![pseudolabel2](https://user-images.githubusercontent.com/47933628/99021064-cd085a00-259a-11eb-802c-21bc643a34de.png)
![pseudolabel3](https://user-images.githubusercontent.com/47933628/99021070-cf6ab400-259a-11eb-8228-4ff38e848568.png)
![esl_thres](https://user-images.githubusercontent.com/47933628/99021074-d396d180-259a-11eb-9d50-c3e24a397d1a.png)

**Comment by Lufei-github** [comment]
The thresholds like this:  
![esl_thres](https://user-images.githubusercontent.com/47933628/99024888-8585cc00-25a2-11eb-8599-d5b29807a3c1.png)

**Comment by Lufei-github** [comment]
And some visualization of the pseudo-labels with both SSL and ESL on a couple of images: the upper left is the SSL method, the upper right is the ESL method, the below is the groundtruth.
![pseudolabel](https://user-images.githubusercontent.com/47933628/99025110-0c3aa900-25a3-11eb-8f10-dc4f3735d71a.png)
![pseudolabel2](https://user-images.githubusercontent.com/47933628/99025116-0f359980-25a3-11eb-8b65-2816994d7642.png)
![pseudolabel3](https://user-images.githubusercontent.com/47933628/99025123-12308a00-25a3-11eb-9c88-be5a21554801.png)

**Comment by Lufei-github** [comment]
Intuitively, I found the ESL method is likely to neglect more pixle, however, these pixle are classified correctly by SSL. Now I am thinking is my 41.31 mIOU's model have poorer performance than your model, or the 0.1 threshold may don't fit the 41.41 mIOU model and I need to pick a more proper threshold. Besides, the all above experiments are trained only with CrossEntropy loss, I didn't add adversarial training into the experiment.

**Comment by SportaXD** [comment]
Your pseudo-labels are fine and your intuition is correct: ESL does neglect more pixel which are deemed less confident in terms of entropy. However, we use in our experiments the pseudo-labels (obtained both from SSL and ESL) as _additional_ supervision (through CrossEntropy loss) on the target domain in the UDA adversarial training which still uses supervision from a source domain. I advice you to use these pseudo-labels in a UDA training if you want to observe improvements in performance.
